cpu/atmega2560/periph/gpio.c

```diff
@@ -53,23 +199,175 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pushpull, gpio_flank_t flank, gpio_cb_t
+
+    switch (dev)
+    {
+#if GPIO_0_EN
+        case GPIO_0:
+            return GPIO_0_PIN & (1 << pin);
+#endif /* GPIO_0_EN */
+
+#if GPIO_1_EN
+        case GPIO_1:
+            return GPIO_1_PIN & (1 << pin);
+#endif /* GPIO_1_EN */
+
+#if GPIO_2_EN
+        case GPIO_2:
+            return GPIO_2_PIN & (1 << pin);
+#endif /* GPIO_2_EN */
+
+#if GPIO_3_EN
+        case GPIO_3:
+            return GPIO_3_PIN & (1 << pin);
+#endif /* GPIO_3_EN */
+
+#if GPIO_4_EN
+        case GPIO_4:
+            return GPIO_4_PIN & (1 << pin);
+#endif /* GPIO_4_EN */
+
+#if GPIO_5_EN
+        case GPIO_5:
+            return GPIO_5_PIN & (1 << pin);
+#endif /* GPIO_5_EN */
+
+#if GPIO_6_EN
+        case GPIO_6:
+            return GPIO_6_PIN & (1 << pin);
+#endif /* GPIO_6_EN */
+    };
 }
 
 void gpio_set(gpio_t dev)
 {
+    uint8_t pin;
+
+    pin = gpio_pin_map[dev];
+
+    switch (dev)
+    {
+#if GPIO_0_EN
+        case GPIO_0:
+            GPIO_0_PORT_REG |= (1 << pin);
+            break;
+#endif /* GPIO_0_EN */
+
+#if GPIO_1_EN
+        case GPIO_1:
+            GPIO_1_PORT_REG |= (1 << pin);
+            break;
+#endif /* GPIO_1_EN */
+
+#if GPIO_2_EN
+        case GPIO_2:
+            GPIO_2_PORT_REG |= (1 << pin);
+            break;
+#endif /* GPIO_2_EN */
+
+#if GPIO_3_EN
+        case GPIO_3:
+            GPIO_3_PORT_REG |= (1 << pin);
+            break;
+#endif /* GPIO_3_EN */
+
+#if GPIO_4_EN
+        case GPIO_4:
+            GPIO_4_PORT_REG |= (1 << pin);
+            break;
+#endif /* GPIO_4_EN */
+
+#if GPIO_5_EN
+        case GPIO_5:
+            GPIO_5_PORT_REG |= (1 << pin);
+            break;
+#endif /* GPIO_5_EN */
+
+#if GPIO_6_EN
+        case GPIO_6:
+            GPIO_6_PORT_REG |= (1 << pin);
+            break;
+#endif /* GPIP_6_EN */
+    };
 }
 
-void  gpio_clear(gpio_t dev)
+void gpio_clear(gpio_t dev)
 {
+    uint8_t pin;
+
+    pin = gpio_pin_map[dev];
+
+    switch (dev)
+    {
+#if GIP_0_EN
+        case GPIO_0:
+            GPIO_0_PORT_REG &= ~(1 << pin);
+            break;
+#endif /*GPIO_0_EN */
+
+#if GPIO_1_EN
+        case GPIO_1:
+            GPIO_1_PORT_REG &= ~(1 << pin);
+            break;
+#endif /*GPIO_1_EN */
+
+#if GPIO_2_EN
+        case GPIO_2:
+            GPIO_2_PORT_REG &= ~(1 << pin);
+            break;
+#endif /*GPIO_2_EN */
+
+#if GPIO_3_EN
+        case GPIO_3:
+            GPIO_3_PORT_REG &= ~(1 << pin);
+            break;
+#endif /*GPIO_3_EN */
+
+#if GPIO_4_EN
+        case GPIO_4:
+            GPIO_4_PORT_REG &= ~(1 << pin);
+            break;
+#endif /*GPIO_4_EN */
+
+#if GPIO_5_EN
+        case GPIO_5:
+            GPIO_5_PORT_REG &= ~(1 << pin);
+            break;
+#endif /*GPIO_5_EN */
+
+#if GPIO_6_EN
+        case GPIO_6:
+            GPIO_6_PORT_REG &= ~(1 << pin);
+            break;
+#endif /* GPIO_6_EN */
+    }
 }
 
 void gpio_toggle(gpio_t dev)
 {
+    if (gpio_read(dev))
+    {
+        gpio_clear(dev);
+    }
+    else
+    {
+        gpio_set(dev);
+    }
 }
```

Comment:
This looks valid. Alternatively the manual saye "*However, writing a logic one to a bit in the PINx Reg-ister, will result in a toggle in the corresponding bit in the Data Register.*. Cou could also try that but don't spend too much time in this